### ISSUE TITLE
Strz type support for UTF-16 and UTF-32

### ISSUE THREAD

**Issue by dreckard** [kaitai-io/kaitai_struct#187]
This issue is very similar to #13 and there is a lot of relevant discussion there.

#### Observed
- If `strz` is used with UTF-16 or UTF-32 then a single null byte is enough to result in termination rather than 2 (U16) or 4 (U32) bytes.

#### Expected
- If the encoding is given as UTF-16 or UTF-32 then the `strz` parse only terminates when the corresponding number of consecutive null bytes appear.

To describe the use case, I am writing a descriptor for the output of an internal Windows tool which reserves a fixed length for the string then null terminates it. While this practice is a bit inefficient and I can't point to any widely used formats that do the same I suspect that it is not uncommon for Windows programs since it is analogous to doing so with UTF-8 in a cross-platform or Unix scenario.

On Windows, an example C struct might look like this:
```
struct TextBlob
{
	UCHAR DataPrefix[n];
	WCHAR WideString[64];
	UCHAR DataSuffix[n];
};
```

For an example data blob and ksy file that that minimally reproduce the issue (at least with the web IDE) please see the following zip:
[utf16_test.zip](https://github.com/kaitai-io/kaitai_struct/files/1088081/utf16_test.zip)

**Comment by GreyCat** [comment]
Ok, to collect all the arguments that were mentioned in previous discussions on this topic in one place:

* Implementing "read until a combination of multiple bytes is encountered" is non-trivial. For example, none of our supported languages have such a method in stdlib (but note almost everyone includes something like "read until a byte is encountered").
* Actually, "read until a sequence of 00 00 is encountered" is plain wrong for this case. Namely, "01 00 00 02 00 00" should be read as two UTF-16LE chars "\u0001\u0200", not as just "\u0001". I.e. "00 00" much appear as whole character, 2-byte aligned.
* It's totally possible (although not really usual) to get this kind of problem outside of C + Windows world. Although majority of applications don't seem to use strings of zero-terminated wide chars family of functions, stuff like `wcslen`, `wcscat`, etc (akin to much more popular `strlen`, `strcat`, etc) technically exists in C since C99 standard.
* No popular (and publically available) format have been found using this so far.
* Looks like this issue is only relevant to UTF-16 representation of strings, and probably 99.9% of the time it is used on actually a fixed string to trim remaining garbage, i.e. it should be solved around [bytes_terminate](https://github.com/kaitai-io/kaitai_struct_cpp_stl_runtime/blob/ed2438847bf2418d67f290679eb2047217d37db3/kaitai/kaitaistream.cpp#L387-L398) function, but probably applied in somewhat different manner.

Current archetypical application generates something like that:

```cpp
foo = bytes_to_str(
    bytes_terminate(
        bytes_strip_right(
            io.read_bytes(20),
            43 // padding byte
        ),
        64, // terminator byte
        false
    ),
    "UTF-8"
);
```

Obviously, applying a function like "bytes_terminate" which operates on strings (not just byte arrays) requires us to convert byte array to string first with "bytes_to_str", i.e. executing something like that:

```cpp
foo = str_terminate(
    bytes_to_str(
        io.read_bytes(20),
    ),
    0x0000, // terminator char
);
```

The big catch is that actually trailing garbage might have something that will be invalid in chosen encoding. Unfortunately, contrary to popular opinion, this is true even for UTF16. For example, if we're working with C++ or PHP (and iconv-based implementation, which converts everything that should be treated as string internally to UTF-8), this string will trigger an error on conversion:

```
61 00|00 00|00 D8 61 00
```

Although we would expect result `a` (as 0x61 is ASCII code for "a", and then string terminates with `00 00`), it will trigger something like InvalidByteSequenceError. This is because `00 D8 61 00` is not a valid UTF16 character: `00 D8` is a lead of surrogate pair, and it must be followed by tail of surrogate pair, and `61 00` is illegal here.

**Comment by sirg3** [comment]
> No popular (and publically available) format have been found using this so far.

[Windows version info resources](https://msdn.microsoft.com/en-us/library/windows/desktop/ms647001(v=vs.85).aspx), found in executables and .res files, use them.

**Comment by GreyCat** [comment]
VS_VERSIONINFO per se has no variable-length strings. However, it includes StringFileInfo → StringTable → [String](https://msdn.microsoft.com/en-us/library/windows/desktop/ms646987(v=vs.85).aspx), which actually includes them.

What's even more peculiar, actually, it seems that there are literally double-null strings to store two-level lists: https://blogs.msdn.microsoft.com/oldnewthing/20091008-00/?p=16443 — I believe they effectively become quad-null terminated strings when laid out in UTF-16. Can anyone confirm/deny that?

**Comment by dreckard** [comment]
Good clarification on the byte alignment requirement. I believe that is correct regarding the quad-null termination for WCHAR string lists. The only other documentation I have seen is in some of the APIs that return data in this style such as [GetLogicalDriveStrings](https://msdn.microsoft.com/en-us/library/windows/desktop/aa364975(v=vs.85).aspx), but I don't see any other way it could be interpreted.

The Windows `KUSER_SHARED_DATA` struct's `NtSystemRoot` is another example of a null terminated UTF-16 (single) string. The struct is fairly commonly parsed in memory dump analysis at least.

```
typedef struct _KUSER_SHARED_DATA
{
     ULONG TickCountLowDeprecated;
     ULONG TickCountMultiplier;
     KSYSTEM_TIME InterruptTime;
     KSYSTEM_TIME SystemTime;
     KSYSTEM_TIME TimeZoneBias;
     WORD ImageNumberLow;
     WORD ImageNumberHigh;
     WCHAR NtSystemRoot[260];
     ULONG MaxStackTraceDepth;
     ...
}
```

http://www.geoffchappell.com/studies/windows/km/ntoskrnl/structs/kuser_shared_data.htm

**Comment by GreyCat** [comment]
I've implemented simple Windows resource parser as an excercise: https://github.com/kaitai-io/windows_resource_file.ksy/blob/master/windows_resource_file.ksy — now it uses a hack to parse strings (especially because there's an extra twist there — same byte space is used to designate numeric and string IDs).

While doing so, I've noticed that there are at least 2 very distinct cases we're talking here:

1. Pure "read until double 00 terminator, convert to string" — encountered at least in NAME and TYPE fields of 
2. "Read fixed number of bytes, then trim it using double 00 terminator logic, then convert to string"

That means my original proposal is wrong. We indeed need both (1) and (2) implemented, not only (2).

**Comment by arekbulski** [comment]
Construct has both (fixed-length string with filler, and c-string). Admittedly it took me 2 years to get it implemented. I have never seen a protocol that actually used the first either, and I cant even imagine why would anyone design such a protocol in the first place, but it was requested on Construct forum more than once, so I guess someone out there actually needs it, so here we are. 

@GreyCat If you give me a go, I will add 2 methods to Python runtime to effectuate this, but you would need to update the compiler (translator). I could update C# then too. Restrition is, these 2 methods need to know what encoding it is, or rather what is the unit size (2 for UTF16, 4 for UTF32, 1 for UTF8).

**Comment by GreyCat** [comment]
Let's start with inventing some ksy syntax that covers all the cases discussed in this ticket.

**Comment by arekbulski** [comment]
Fixed string: "type: str, size: N, encoding: utf16, [unitsize: T=2]"

Reads N bytes, then successively strips last T bytes if they are null, down to empty string. If encoding is recognizable like UTF*, unitsize can be inferred. N must be multiple of T.

CString: "type: str, terminator: 0, encoding: utf16, [unitsize: T=2]"
CString: "type: strz, encoding: utf16, [unitsize: T=2]"

Reads T bytes at a time, until that chunk is all null bytes. If first chunk is nulls, its an empty string. If encoding is recognizable like UTF*, unitsize can be inferred. Terminator other than 0 should be compile error, because at least UTF encodings support only one way of terminating it.

**Comment by arekbulski** [comment]
By recognizable encodings I mean those:

construct.possiblestringencodings = {'U16': 2, 'utf_8': 1, 'utf32': 4, 'utf_32_le': 4, 'utf8': 1, 'utf_32_be': 4, 'StringsAsBytes': 1, 'utf_32': 4, 'utf_16_be': 2, 'U32': 4, 'utf16': 2, 'ascii': 1, 'utf_16': 2, 'utf_16_le': 2, 'U8': 1}¶

**Comment by arekbulski** [comment]
There is also a problem when both _size_ and _terminator_ are used.
```py
meta:
  id: test1
seq:
  - id: value
    type: str
    terminator: 0
    size: 10
    encoding: utf-8
```
Compiles into following. Problem is, _bytes_terminate_ only supports single characters.
```py
(KaitaiStream.bytes_terminate(self._io.read_bytes(10), 0, False)).decode(u"utf-8")
```
Extract from the runtime:
```py
    def bytes_terminate(data, term, include_term):
        new_len = 0
        max_len = len(data)
        while new_len < max_len and data[new_len] != term:  #<--- indexing not slicing
            new_len += 1   #<--- non variable
        if include_term and new_len < max_len:
            new_len += 1
        return data[:new_len]
```

**Comment by GreyCat** [comment]
`bytes_terminate`, as its name suggest does not support *any* characters at all, only bytes. What problem do you see here?

**Comment by arekbulski** [comment]
To support UTF16/32 for example, bytes_terminate would need to slice `data[newlen:newlen+unit]` and then increment like `newlen += unit`. In other words, current implementation would not support UTF16/32. Thats what this issue is about. And by characters I meant single bytes, sorry.

**Comment by GreyCat** [comment]
It doesn't, that's true. My main concern here is that technically we should *not* deal with *bytes* at all: if we're dealing with encodings like UTF16, for example, in C++, that would call for `char16_t` type instead of `uint8_t` and relevant reinterpret cast.

If we'll stick with bytes-centric implementation, however, from runtime's point of view, we can probably cover all possible cases by specifying something like `byte_terminate_multi`, something like that:

```java
byte[] bytesTerminateMulti(byte[] bytes, byte[] term, int unitSize, boolean includeTerm)
```

and the same thing about `bytes_strip_right_multi`:

```java
public static byte[] bytesStripRight(byte[] bytes, int unitSize, byte[] padBytes) {
```

**Comment by arekbulski** [comment]
Good idea, it would be better to have a separate (multi) method instead of generalizing the existing (single) method, because multi will have less performance than single. 

Would _unitsize_ parameter be actually needed? I think its just _len(term)_ and _len(padbytes)._

**Comment by GreyCat** [comment]
Specifying `unitSize` is need because we need so support both aligned `00 00` and unaligned variants, i.e. something like: `30 00 40 00 50 50 50 00 00 12 34 56` should `30 00 40 05 50 50 50` (or `O\0@\0P` in ASCII) after termination, if `unitSize` is 1 and `term` is `[0, 0]`.

Also, I think that this place is more than anything warrants plenty of language-specific APIs. For example, in C/C++, you don't want to pass an array around, you'd want one `uint16_t` integer and then one just `reinterpret_cast`s parts of the array and compares them to that integer. This obviously doesn't make much sense for higher-level languages, in which "parsing 2 bytes as integer" is an expensive operation.

**Comment by arekbulski** [comment]
I thought it should be only unit-aligned version. The usecase for aligned would be obviously strings, but what would be the usecase for unaligned?

**Comment by KOLANICH** [comment]
BTW, do we really have to carve `strz`s ourselves for C++? I mean that string functions, like basic_string ctor from a pointer, expect certain termination, so can't we just pass a pointer, and then shift to the ```size()``` of the string parsed by c++?

**Comment by arekbulski** [comment]
It probably only works with single-null terminator, right? This issue is about supporting UTF16 and 32.

**Comment by KOLANICH** [comment]
```wchar_t``` is a utf-16 encoded char. Wstring is a string of ```wchar_t```s. Arrays of wchar_t are assummed to be terminated with 2 null bytes.

**Comment by arekbulski** [comment]
So someone uses it with UTF32 encoding, then what?

**Comment by GreyCat** [comment]
> `wchar_t` is usually (though it is not standardised) a utf-16 encoded code point.

I'd say that from practical point of view, the only system that uses UTF16 wide chars is Windows. Virtually everyone else use UTF32 there.

**Comment by arekbulski** [comment]
Python unicode strings use 1/2/4 bytes per character, depending on actual text.

**Comment by GreyCat** [comment]
The question @KOLANICH raised was about `wchar_t` in C++. Python, Java, C#, Go and everything else is non-relevant to that.

Let's get back to original topic.

> The usecase for aligned would be obviously strings, but what would be the usecase for unaligned?

I believe someone further above this issue provide some examples why that would be useful. I recall some strings in UTF16 wanted 4-byte [0, 0, 0, 0] terminator and stuff like that.

**Comment by kdumontnu** [comment]
Any update on this issue? Seems like there is a path forward. Need any help? 

This is currently causing me some grief.

**Comment by ricky-daniel13** [comment]
![image](https://github.com/kaitai-io/kaitai_struct/assets/47061347/fc0ce5a5-2614-43f1-97df-2614826265ed)

It is definetly a more common ocurrence than what it is assumed. Trying to make a struct for the subtitle files of a game, which are UTF-16 strings that end on a terminator.

**Comment by AtomCrafty** [comment]
This is hardly a good solution, but it is possible to parse zero terminated utf-16 (and utf-32) strings by abusing the `repeat-until` feature.

The basic idea is to just parse it as an array of `u2`/`u4` with a `repeat-until` condition of `_ == 0`. After that you can read the length of that array to parse the actual string in an `instance`. This trick does however require you to be able to write an expression to calculate the start address, which can become tricky if the location depends on many factors.

```yaml
seq:
  - id: before_string
    type: u4
  - id: string_chars
    type: u2
    repeat: until
    repeat-until: _ == 0
  - id: after_string
    type: u4
instances:
  my_string:
    pos: 0x04
    type: str
    encoding: utf-16
    size: _root.string_chars.size * 2
```

On the issue of relevance: I work with a lot of obscure game engines and zero terminated utf-16 strings are not that uncommon there, for example for the file name tables in their archive formats.

**Comment by generalmimon** [comment]
@AtomCrafty:

> This is hardly a good solution, but it is possible to parse zero terminated utf-16 (and utf-32) strings by abusing the `repeat-until` feature.

No, I think this is actually quite a good workaround (and I don't think this is "abusing", it reflects well how a null-terminated UTF-16 string is parsed and it's as close to a "native" solution as you can get). It can be also extracted to a reusable type like this:

```ksy
meta:
  id: strz_utf_16_test
  endian: le
seq:
  - id: before_string
    type: u4
  - id: my_string
    type: strz_utf_16
  - id: after_string
    type: u4
types:
  strz_utf_16:
    seq:
      - id: value
        size: 2 * (code_units.size - 1)
        type: str
        encoding: utf-16le
      - id: term
        type: u2
        valid: 0
    instances:
      code_units:
        pos: _io.pos
        type: u2
        repeat: until
        repeat-until: _ == 0
```

Parsing the string in `seq` and performing the lookahead in `instances` is a bit more [serialization](https://doc.kaitai.io/serialization.html)-friendly. You can set the `value` string normally (though it must not contain any U+0000 code points particularly if it comes from a user, otherwise it won't be read back properly), ensure that the `code_units` list has a correct length (the contents won't matter) and then disable writing of the `code_units` instance (see https://doc.kaitai.io/serialization.html#_parse_instances).

<details>
<summary>Python code to handle serialization of the <code>strz_utf_16_test.ksy</code> spec above</summary>

```py
from kaitaistruct import KaitaiStream
from strz_utf_16_test import StrzUtf16Test

r = StrzUtf16Test()
r.before_string = 803_200_000

my_s = StrzUtf16Test.StrzUtf16(None, r, r._root)
my_s.value = "Hello \U0001F44B"
my_s.term = 0
my_s.code_units = [0x0000] * (len(my_s.value.encode('utf-16le')) // 2 + 1)
my_s.code_units__to_write = False
my_s._check()

r.my_string = my_s
r.after_string = 48_160_000
r._check()

import io
_io = KaitaiStream(io.BytesIO(bytearray(4 + 2 * len(r.my_string.code_units) + 4)))
r._write(_io)

output = _io.to_byte_array()
print(output.hex(' '))  # 00 dc df 2f 48 00 65 00 6c 00 6c 00 6f 00 20 00 3d d8 4b dc 00 00 00 dd de 02
```

</details>

---

> This trick does however require you to be able to write an expression to calculate the start address, which can become tricky if the string is located behind other variable-size structures or within a type definition.

This requirement can be eliminated using `_io.pos` (as you can see in the above .ksy snippet). Since the `code_units` instance is invoked just before `value` is parsed, the `_io.pos` expression gives you the current position relative to the current stream that can be used in `pos`. Also, `instances` are cached, so even if `code_units` is requested later (when `_io.pos` would probably give a different value), it's not parsed again - the cached value from the first invocation is simply returned.

**Comment by AtomCrafty** [comment]
@generalmimon 

Well you seem to know you way around Kaitai way better than me, so I'll take your word for it being idiomatic ^^
I did not know about `_io.pos` or that instances can be evaluated before the main sequence is finished.
Your variation makes the whole thing look much cleaner than what I cobbled together and it has the major advantage of being plug-and-play. I'm pretty happy with that as a solution.

**Comment by t0xicCode** [comment]
Another implementation I've used in the past to get UTF-16 support. Using the `as_string` instance does the conversion from UTF-16LE to the native encoding.

```yaml
unicode_16:
  seq:
    - id: first
      size: 0
      if: start_ >= 0
    - id: c
      type: u2
      repeat: until
      repeat-until: _ == 0
    - id: last
      size: 0
      if: end_ >= 0
  instances:
    start_:
      value: _io.pos
    end_:
      value: _io.pos
    as_string:
      pos: start_
      type: str
      size: end_ - start_ - 2
      encoding: UTF-16LE
```

**Comment by generalmimon** [comment]
I'm still not sure about the .ksy syntax, even though there was some proposal by @arekbulski in https://github.com/kaitai-io/kaitai_struct/issues/187#issuecomment-364717635. I have a few remarks:

> Terminator other than 0 should be compile error, because at least UTF encodings support only one way of terminating it.

I don't think this is great in terms of language design. And I don't like repurposing the syntax `terminator: 0` for multi-byte zero terminators at all, given that `terminator` only accepts byte values at the moment and its valid range is `0-255`. If we admitted that `terminator` with an integer value can specify a multi-byte terminator sequence, it wouldn't be clear from the integer value how long the terminator is.

And the `0` in `terminator: 0` doesn't feel like a special value that cannot be changed arbitrarily. However, if we ever wanted to allow non-zero terminators, anything other than `terminator: 0` (e.g. `terminator: 0x20`) would be ambiguous due to unclear endianness.

---

I think it's better to extend `terminator` to not only accept integers (as byte values - `terminator: 0-255`), but also byte arrays (as more or less suggested in https://github.com/kaitai-io/kaitai_struct/issues/158). For a null-terminated UTF-16, one could use `terminator: [0, 0]` + `type: str` + `encoding: UTF-16LE`. The upside is that the syntax is unambiguous (except for the "unit size", which... well, must be explained in the documentation), the downside is that it is more verbose than just `terminator: 0`, but that can be solved by making `type: strz` + `encoding: UTF-16LE` an alias of that. That said, `type: strz` will no longer be an equivalent of `type: str` + `terminator: 0`, but rather `type: str` + `terminator: [0, ..., 0]`, where the size of `terminator` varies depending on the `encoding`.

This means that `type: strz` + `encoding: UTF-{16,32}*` will now work as most people expect.

For encodings unknown to the compiler, it seems fine to assume 1 zero byte as a terminator for `type: strz`, since there are very few encodings where this doesn't work (I don't know of any other than UTF-16 and UTF-32).

`terminator: 0` + `type: str` + `encoding: UTF-{16,32}*` would remain single null byte terminated (since `terminator: 0` would be basically the same as `terminator: [0]`). This is most likely not what the user intended, though, so it would be good to emit a warning suggesting to use `type: strz` without `terminator` (or `type: str` + `terminator: [0, 0] / [0, 0, 0, 0]`) instead. Even without this warning, the user would find out very soon that it isn't working properly, but the warning would advise how to fix it.

@GreyCat:

https://github.com/kaitai-io/kaitai_struct/issues/187#issuecomment-371177465

> > The usecase for aligned would be obviously strings, but what would be the usecase for unaligned?
>
> I believe someone further above this issue provide some examples why that would be useful. I recall some strings in UTF16 wanted 4-byte [0, 0, 0, 0] terminator and stuff like that.

I don't think unaligned terminators are ever used for strings, and I don't think any of the discussion above mentioned a real-world example. I haven't found any evidence of this on the internet either (both UTF-16 and UTF-32 seem to be universally null-terminated by a group of 2/4 zero bytes only at an offset that is a multiple of 2/4).

I think it makes sense if you consider that the parsing of a null-terminated UTF-16/UTF-32 string would be implemented in C/C++ by treating the string as an array of `wchar_t` and searching for the element equal to `0`. Searching for an *unaligned* group of zero bytes would be slower, arguably more complicated to implement and it wouldn't really make sense (as @GreyCat pointed out in https://github.com/kaitai-io/kaitai_struct/issues/187#issuecomment-309768854, looking for an unaligned sequence of `00 00` is wrong for UTF-16, because a string of 2 chars `"\u0001\u0200"` as `01 00 00 02` in UTF-16LE).

Unaligned terminators might still be useful for something other than strings, though (e.g. byte arrays, which `terminator` also supports). Seemingly some real-world use case in JPEG is given in https://github.com/kaitai-io/kaitai_struct/issues/158, which appears to use an unaligned multi-byte terminator. However, such general "scanning" scenarios probably fall more into https://github.com/kaitai-io/kaitai_struct/issues/538.

Regardless of whether the "unit size" is directly configurable from the .ksy or not, it seems that the length of `terminator` is a good default.

**Comment by generalmimon** [comment]
> Regardless of whether the "unit size" is directly configurable from the .ksy or not, it seems that the length of `terminator` is a good default.

Hmm, I'm starting to think that this would be a bad idea. I think this behavior might be quite confusing - consider this:

```ksy
seq:
  - id: foo
    terminator: [0x00, 0x00]
```

If we were to ask someone what they think this does (let's assume they already know the type is a byte array, since there's no `type`), the most obvious answer would be "it looks for 2 consecutive zero bytes and that will be the end of `foo`". And it's hard to reasonably justify why this intuitive interpretation is wrong if the above .ksy snippet actually only checked for the terminator at an even offset (the intuitive interpretation would match if the default was `unit-size: 1`, not `unit-size: 2`). Whether or not we wrote about this pitfall in the documentation wouldn't change the fact that this would be counter-intuitive.

Perhaps we should have `unit-size: 1` as the default after all. This would ensure that `terminator: [0x00, 0x00]` would match two zero bytes regardless of what offset they are at. For the use case of UTF-16/UTF-32 strings, `type: strz` would be strongly preferred, which would automatically select the appropriate null terminator for the particular `encoding`:

```ksy
seq:
  - id: foo
    type: strz
    encoding: UTF-16LE
```

The long equivalent would be as follows:

```ksy
seq:
  - id: foo
    unit-size: 2
    terminator: [0x00, 0x00]
    type: str
    encoding: UTF-16LE
```

@GreyCat What do you think?

**Comment by GreyCat** [comment]
My gut feeling is that eventually we'd want to converge on implementation akin to #538. Everything we do regarding string terminations — e.g. `strz`, `terminator`, etc, is just a shortcut to one of the possible options in that engine.

Scanning engines can be provided for seeking 1 byte with specific pattern, 2 bytes with specific pattern unaligned/aligned, 4 bytes, etc, any kind of combinations/permutations.

Logically, from standpoint of least possible friction, having some syntactic sugar which will make `strz` behave as you've explained, absolutely seem to make sense to me. That's what 99% of users will need and that's what C/C++ implementations with UTF16 null-terminated strings do.

**Comment by generalmimon** [comment]
> My gut feeling is that eventually we'd want to converge on implementation akin to #538. Everything we do regarding string terminations — e.g. `strz`, `terminator`, etc, is just a shortcut to one of the possible options in that engine.

> Logically, from standpoint of least possible friction, having some syntactic sugar which will make `strz` behave as you've explained, absolutely seem to make sense to me.

Makes sense, thanks for commenting. I'll probably keep it simple then and only implement `type: strz` + `encoding: UTF-{16,32}*` for now, with no other options. As you mentioned, this is the what most people really need.

Anything more complex will be the subject of https://github.com/kaitai-io/kaitai_struct/issues/538.

To be clear, does that mean that we will not pursue https://github.com/kaitai-io/kaitai_struct/issues/158, and will instead move that to https://github.com/kaitai-io/kaitai_struct/issues/538 as well? If so, perhaps we should close https://github.com/kaitai-io/kaitai_struct/issues/158 in favor of https://github.com/kaitai-io/kaitai_struct/issues/538 for clarity.

**Comment by GreyCat** [comment]
> To be clear, does that mean that we will not pursue https://github.com/kaitai-io/kaitai_struct/issues/158, and will instead move that to https://github.com/kaitai-io/kaitai_struct/issues/538 as well? If so, perhaps we should close https://github.com/kaitai-io/kaitai_struct/issues/158 in favor of https://github.com/kaitai-io/kaitai_struct/issues/538 for clarity.

Generally, I would agree. I believe that design with "scanners" concept (#538) in general is solid, it's just a question of careful implementation.

> Wildcard bytes, regular expressions, number ranges and other helpers could also be of assistance in defining terminators in other file formats.

All these things absolutely can be covered with custom scanning procedures. We can do a library of built-in scanning procedures in different languages, that's for sure, but ultimately it will be to up user to pick and use one that fits their purpose.

**Comment by generalmimon** [comment]
This is now implemented in all 11 target languages that Kaitai Struct supports (C++/STL, C#, Go, Java, JavaScript, Lua, Nim, Perl, PHP, Python, Ruby).

Unfortunately, it cannot be used in Lua yet, because our Lua runtime library doesn't support UTF-16* or UTF-32* encodings at all, but it has been implemented there as well, so it will be available once the UTF-{16,32}* encodings are implemented (see https://github.com/kaitai-io/kaitai_struct_compiler/commit/b0cbf6d2f7b60909b0e50ae33ae22fec1a8a22a3).

The corresponding tests at https://ci.kaitai.io/ are called `StrPadTermUtf16` (this is for the `size` + `type: strz` case) and `TermStrzUtf16V{1,2,3,4}` (for parsing `type: strz` directly from the stream, without any known `size`). They don't pass in Lua as expected (`Encoding UTF-16LE not supported`).

The fact that they don't pass in Nim at the moment is an infrastructure issue that will disappear after the Docker image for Nim gets updated (by the CI pipeline at https://github.com/kaitai-io/kaitai_struct_docker_images). Unlike all other languages, the Nim Docker image bundles the runtime library inside ([`src/nim/1.6.0-linux-x86_64/Dockerfile:7`](https://github.com/kaitai-io/kaitai_struct_docker_images/blob/185125a7f4a52540424b7a0e18ad904be5fd452e/src/nim/1.6.0-linux-x86_64/Dockerfile#L7)) instead of using the latest runtime library mapped at [`/runtime` in the Docker container](https://github.com/kaitai-io/kaitai_struct_tests/blob/35a01047c703a80c10203ccb38aad8e085c762c3/docker-ci#L230). As a result, it uses the old version of the Nim runtime library that doesn't implement the necessary methods.

I also ran into this problem when testing locally and fixed it by adding `( cd "$NIM_RUNTIME_DIR" && nimble develop -y )` at [`ci-nim:7`](https://github.com/kaitai-io/kaitai_struct_tests/blob/35a01047c703a80c10203ccb38aad8e085c762c3/ci-nim#L7). I don't think `ci-nim` is the right place for resolving this properly, though. Instead, this should ideally be performed when building the Docker image.

When tested locally, the new tests passed in Nim.
